### ISSUE TITLE
Improve track prediction Kalman filter for PTZ targeting

### ISSUE THREAD

**Issue by Micky-G** [bobcamera/bobcamera#45]
We need to improve the Kalman filter used for our track prediction. Maybe use the version in the SORT implementation as reference. We will need this in order to help predict the path of the object being tracked so that we can try and direct the PTZ to the target object.

**Comment by jdmaudio** [comment]
Track predication in the SORT Kalman filter is working well

**Comment by jdmaudio** [comment]
Yes we can access predicted centre points for any tracks from SORT via:

```
std::map<int, Track> SORT::Tracker::GetTracks() const {
    return tracks_;
}

then call...

const std::vector<cv::Point>& Track::get_predictor_center_points() const 
{
    return predictor_center_points_;
}
```

Eventually maybe it would be good for us to model some scenarios for PTZ tracking?  E.g. for a camera with a certain capability in terms steering and zoom, can we reliably follow the predicted centre point of an object with x velocity in pixel space?

**Comment by jdmaudio** [comment]
Some further notes on Kalman filtering.

The Kalman Filter we use currently is tuned fairly well for all-sky cam tracking purposes. The filter employs a constant velocity model with some basic adaptive filtering that adjusts process noise when any object manoeuvring is occurring (e.g. acceleration or change of direction). We adapt based on the error residual - when the residual gets large we increase process noise which causes the filter to favour incoming measurements over the prediction from the filter. When residual is low it decreases process noise. 

Whilst this works well, more sophisticated multiple model filtering might be worth exploring in future, and/or with use of Unscented Kalman Filter which can account for non linearities.

**Comment by jdmaudio** [comment]
The following library is recommended for other filtering options:
https://github.com/robotology/bayes-filters-lib
